Embedded 3D Previewer

benfen opened the pull request:
Resolves https://github.com/NREL/floorspace.js/issues/282

Port logic from https://github.com/macumber/emscripten-test to add a 3D previewer to floorspace.js.  Adds an option to the `Save Floorplan` button to export to threeJS JSON instead; also adds a button to the top left to directly open the current floorspace in three.js.

The majority of the cpp files here are pulled from openstudio.

The PR is deployed at: https://benfen.github.io/floorspace.js/

benfen committed:
Tweak save as modal to support different export types

benfen committed:
Add capability to downloda threeJS json of floorplan

benfen committed:
Tweak filepaths for items in the `/static` path

benfen committed:
Add 3D Viewer for HTML

benfen committed:
Add source for wasm for generating threeJS json

benfen committed:
Tweak how boost is added as a dependency

benfen committed:
Fix e2e tests

benfen committed:
Add e2e test for 3D export

benfen committed:
Clean up cmake file

benfen committed:
Add flag to toggle display of the 3D preview button

benfen committed:
Extend feature flag to cover export options

macumber commented:
It looks like the WASM doesn't end up in the `embeddable_geometry_editor.html` which the OS App uses which is good for me.  I'll check that the embedded build still works in a day or two.

benfen committed:
Update build system to use webpack for 3DViewer

The webpage for the 3D viewer is now connected to the same webpack configuration as the rest of the application.

Reworked a fair amount of the js for the 3D viewer to make it more modern and fix some scoping issues.

bgschiller commented on src/App.vue:
```diff
@@ -86,6 +86,10 @@ export default {
+
+    const threeJSModuleScript = document.createElement('script');
+    threeJSModuleScript.setAttribute('src', '3DViewer/libtest_lib.js');
+    document.head.appendChild(threeJSModuleScript);
```
Is your goal to load this async? You could also use `<script src="3DViewer/libtest_lib.js" async></script>` in the index.html file

benfen replied:
I had some issue where adding a `<script>` tag caused the build to not inject the other scripts.  I'm not sure why it happened, but it's not happening any more, so I'll update that.

benfen replied:
I discovered some new and exciting issues when using a `<script>` tag - namely, the build fails because something in webpack is trying to modify the `src` attribute.

I'm going to stick with the JS approach for now.

bgschiller commented on src/components/Toolbar.vue:
```diff
@@ -218,10 +218,13 @@ export default {
     },
+    open3DPreviewer() {
+      localStorage.setItem("floorplan3DExport", JSON.stringify(application.$store.getters['exportData']));
+      window.open('3DViewer');
```
this seems to be plenty fast, but if you wanted to avoid localstorage the postMessage api might be another option.

benfen replied:
In this case, I do want it to be stored in case the user refreshes the tab or reopens it.

bgschiller reviewed (approved):
Super cool, thanks for tackling this

benfen committed:
Add script tag to html instead of loading it dynamically in the vue

macumber commented:
`npm run openstudio-build` is failing for me after these latest commits

```
macumber@Automatic-Magic-PC MINGW64 /c/repos/floorspace.js (feat/embedded-previewer)
$ npm run openstudio-build

> floorspace.js@0.7.0 openstudio-build C:\repos\floorspace.js
> node build/build.js && node build/build-embeddable.js

  Tip:
  Built files are meant to be served over an HTTP server.
  Opening index.html over file:// won't work.

cp: no such file or directory: static/*
Hash: 845ba641053087b2cfb2
Version: webpack 1.15.0
Time: 25227ms
                                                     Asset       Size  Chunks             Chunk Names
                    static/fonts/element-icons.6f0a763.ttf      11 kB          [emitted]
                static/js/manifest.9fa52e0d6e84e9c8767c.js  902 bytes       0  [emitted]  manifest
                  static/js/vendor.c53abce85da2e193422d.js    1.75 MB    1, 0  [emitted]  vendor
                     static/js/app.77b89bc4e5afd7b432ae.js     286 kB    2, 0  [emitted]  app
                  static/js/viewer.72749e8e1a4421521267.js    20.5 kB    3, 0  [emitted]  viewer
       static/css/app.e034f7014dba0322645c4385deb681d3.css     224 kB    2, 0  [emitted]  app
    static/css/viewer.fd865ab4ed6f1814413a2a2a3b80d400.css  859 bytes    3, 0  [emitted]  viewer
            static/js/manifest.9fa52e0d6e84e9c8767c.js.map    9.21 kB       0  [emitted]  manifest
              static/js/vendor.c53abce85da2e193422d.js.map    10.8 MB    1, 0  [emitted]  vendor
                 static/js/app.77b89bc4e5afd7b432ae.js.map     1.5 MB    2, 0  [emitted]  app
   static/css/app.e034f7014dba0322645c4385deb681d3.css.map     297 kB    2, 0  [emitted]  app
              static/js/viewer.72749e8e1a4421521267.js.map     115 kB    3, 0  [emitted]  viewer
static/css/viewer.fd865ab4ed6f1814413a2a2a3b80d400.css.map    1.69 kB    3, 0  [emitted]  viewer
                                                index.html    1.11 kB          [emitted]
                                       3DViewer\index.html    1.27 kB          [emitted]
Loading scripts from: 3DViewer/libtest_lib.js


internal/fs/utils.js:269
    throw err;
    ^

Error: ENOENT: no such file or directory, open './dist3DViewer/libtest_lib.js'
    at Object.openSync (fs.js:462:3)
    at readFileSync (fs.js:364:35)
    at C:\repos\floorspace.js\build\build-embeddable.js:10:10
    at Array.map (<anonymous>)
    at Object.<anonymous> (C:\repos\floorspace.js\build\build-embeddable.js:7:55)
    at Module._compile (internal/modules/cjs/loader.js:999:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
    at Module.load (internal/modules/cjs/loader.js:863:32)
    at Function.Module._load (internal/modules/cjs/loader.js:708:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:60:12) {
  errno: -4058,
  syscall: 'open',
  code: 'ENOENT',
  path: './dist3DViewer/libtest_lib.js'
}
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! floorspace.js@0.7.0 openstudio-build: `node build/build.js && node build/build-embeddable.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the floorspace.js@0.7.0 openstudio-build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\macumber\AppData\Roaming\npm-cache\_logs\2021-08-05T18_39_46_486Z-debug.log
```

[2021-08-05T18_39_46_486Z-debug.log](https://github.com/NREL/floorspace.js/files/6940973/2021-08-05T18_39_46_486Z-debug.log)

macumber commented:
I'll try to add your changes to the previewer to https://github.com/NREL/openstudio-common-measures-gem/pull/81

benfen committed:
Fix build issue with the wasm script

benfen commented:
Had some weird issues with the build not liking a `<script>` tag.  It's back to working again.

benfen committed:
Fix how the wasm script is imported

bgschiller reviewed (approved):
Is it possible to keep the 3DViewer/build directory out of version control?

benfen commented:
@bgschiller It's definitely possible - I added it to source control since I don't expect it to change frequently in comparison to the JS source code and I didn't want to add that as an additional build step/dependency.
